pkg/api/customization/podsecuritypolicytemplate/podsecuritypolicytemplate.go

```diff
@@ -0,0 +1,88 @@
+package podsecuritypolicytemplate
+
+import (
+	"fmt"
+
+	"github.com/rancher/norman/types"
+	"github.com/rancher/types/apis/management.cattle.io/v3/schema"
+	"github.com/rancher/types/client/management/v3"
+	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/api/errors"
+)
+
+type Store struct {
+	types.Store
+}
+
+func (s *Store) Delete(apiContext *types.APIContext, schema *types.Schema, id string) (map[string]interface{}, error) {
+	projectHasPSPT, err := projectHasPSPTAssigned(apiContext)
+	if err != nil {
+		return nil, fmt.Errorf("error checking if PSPT is assigned to projects: %v", err)
+	}
+
+	if projectHasPSPT {
+		return nil, errors.NewBadRequest("PSPT is assigned to one or more projects, remove PSPT from those " +
+			"projects before deleting")
+	}
+
+	clusterHasPSPT, err := clusterHasPSPTAssigned(apiContext)
+	if err != nil {
+		return nil, fmt.Errorf("error checking if PSPT is assigned to clusters: %v", err)
+	}
+
+	if clusterHasPSPT {
+		return nil, errors.NewBadRequest("PSPT is assigned to one or more clusters, remove PSPT from those " +
+			"clusters before deleting")
+	}
+
+	return s.Store.Delete(apiContext, schema, id)
+}
+
+func Formatter(apiContext *types.APIContext, resource *types.RawResource) {
+	// check if PSPT is assigned to a cluster or project
+	projectHasPSPT, err := projectHasPSPTAssigned(apiContext)
+	if err != nil {
+		logrus.Warn("failed to determine if PSPT was assigned to a project: %v", err)
+		return
+	}
+
+	if projectHasPSPT {
+		// remove delete link
+		delete(resource.Links, "remove")
+		return
+	}
+
+	clusterHasPSPT, err := clusterHasPSPTAssigned(apiContext)
+	if err != nil {
+		logrus.Warnf("failed to determine if a PSPT was assigned to a cluster: %v", err)
+		return
+	}
+
+	if clusterHasPSPT {
+		// remove delete link
+		delete(resource.Links, "remove")
+		return
+	}
+}
+
+func projectHasPSPTAssigned(apiContext *types.APIContext) (bool, error) {
+	projectSchema := apiContext.Schemas.Schema(&schema.Version, client.ProjectType)
+	projects, err := projectSchema.Store.List(apiContext, projectSchema, &types.QueryOptions{
```

Comment:
This is going to be too expensive. It adds an extra API request to k8s for every pspt every time it is listed. I think we need to switch to a cache (aka informer or indexer). Same deal for clsuterHasPSPTAssigned.

Reply:
However, I think the validation you do on delete SHOULD NOT use the cache. It should make a real api request just to be sure.

Reply:
Added indexers for project/cluster for use only in the Formatter func